components/server/src/user/user-controller.ts

```diff
@@ -281,6 +286,95 @@ export class UserController {
                 res.sendStatus(200);
             },
         );
+
+        router.post(
+            "/auth/sendHeartbeat/:instanceID",
+            async (req: express.Request, res: express.Response, next: express.NextFunction) => {
+                if (!req.isAuthenticated() || !User.is(req.user)) {
+                    res.sendStatus(401);
+                    log.warn("unauthenticated send heartbeat", { instanceId: req.params.instanceID });
+                    return;
+                }
+
+                const user = req.user as User;
+                if (user.blocked) {
+                    res.sendStatus(403);
+                    log.warn("blocked user attempted to send heartbeat", {
+                        instanceId: req.params.instanceID,
+                        userId: user.id,
+                    });
+                    return;
+                }
+
+                const wasClosed = req.body.wasClosed;
+                // const roundTripTime = req.body.roundTripTime;
+                const instanceID = req.params.instanceID;
+                if (!instanceID) {
+                    res.sendStatus(400);
+                    log.warn("attempted to send heartbeat without instance ID", {
+                        instanceId: req.params.instanceID,
+                        userId: user.id,
+                    });
+                    return;
+                }
+                const [workspace, instance] = await Promise.all([
+                    this.workspaceDB.findByInstanceId(instanceID),
+                    this.workspaceDB.findInstanceById(instanceID),
+                ]);
+                if (!workspace || instance?.status.phase !== "running") {
+                    res.sendStatus(404);
+                    log.warn("attempted to send heartbeat for non-existent workspace instance", {
```

Comment:
should this be `log.warn(<context>, <message>)`? [[1]](https://github.com/gitpod-io/gitpod/blob/main/components/gitpod-protocol/src/util/logging.ts#L53)

I think `userId`, `istanceId` are part of [LogContext](https://github.com/gitpod-io/gitpod/blob/main/components/gitpod-protocol/src/util/logging.ts#L14-L19)

Reply:
@andreafalzetti  See here https://github.com/gitpod-io/gitpod/blob/9b46ff3a216d5e598b91abf832569b91bad3e10d/components/server/src/user/user-controller.ts#L205-L208

Reply:
I thought `log` support a lot of param pairs, and we used to make log string first param

Reply:
I'm not quite sure if the order of params will affect the query in gcloud, I unresolved this conversation and maybe webApp team can answer us

Reply:
> I'm not quite sure if the order of params will affect the query in gcloud

@mustard-mh It does indeed.

> we used to make log string as first param in most cases

That's not good. The expected order is: logContext, message, error, payload.